ENCD-6157-allow-multiple-types-for-report-view-button

forresttanaka opened the pull request:
* The view control buttons now have a very simple plug-in mechanism that uses a new entry in `viewTypeMap` to determine which component should render each of the three types of view control buttons (search, report, matrix).
* Each of the three types of buttons have their own renderers now so they can easily have more custom actions without cramming them into a single component and doing type checking.

~https://encd-6157-mixed-reports-pv1-fytanaka.demo.encodedcc.org/~

forresttanaka committed:
ENCD-6157-mixed-reports

emro commented on src/encoded/static/components/__tests__/search-controls.js:
```diff
@@ -37,4 +38,43 @@ describe('View controls for search with Experiments', () => {
         });
     });
+
+    describe('Mutliple searched types', () => {
```
typo: "Multiple"

emro commented on src/encoded/static/scss/encoded/modules/_search.scss:
```diff
@@ -370,6 +370,40 @@ $form-bg: #f3f3f3;
+        white-space: nowrap;
+
+        a {
+            font-family: Quicksand, sans-serif;
```
I think you could use $button-font here, just in case we one day wanted to change all the instances of Quicksand - unlikely now but theoretically possible.

emro commented on src/encoded/static/components/view_controls.js:
```diff
@@ -1,24 +1,160 @@
+                {viewType.title}
+            </ViewControlButton>
+        );
+    }
```
It looks like you could use the `<SearchViewButton>` here in lines 66-68 but I think you didn't do that because the code is easier to understand if you have one component for each type of view button. You could rename the SearchViewButton to be something more generic and use it here, but I also think it's fine as-is, it was just a thought I had.

forresttanaka replied:
Oh, that's a good idea.

emro reviewed (commented):
Looks good! Just a couple comments, but no changes needed.

forresttanaka committed:
ENCD-6157-mixed-reports

Address code-review feedback

forresttanaka committed:
ENCD-6157-mixed-reports

Fix subtle visual issue when the last button within btn-attached has a dropdown. The dropdown `<ul>` causes the :last-child to not apply. Now I use :last-of-type.

forresttanaka commented:
I made your suggested modifications, and then noticed a one-pixel issue when the last (or only) button within .btn-attached has a dropdown because of the added `<ul>`, so I changed the :last-child into :last-of-type to target just the button style.

forresttanaka commented:
Updated demo at: ~https://encd-6157-mixed-reports-pv2-fytanaka.demo.encodedcc.org/~

emro reviewed (approved):
Looks good!
